### PR TITLE
doc: Revert back to using breathe instead of docleaf

### LINKED ISSUE #61284: Docleaf tool is using a non-OSI approved license
**Describe the bug**

#59570 introduced Docleaf as a replacement for Breathe to make Doxygen documentation accessible from Sphinx.
As opposed to Breathe which was BSD licensed, Docleaf is published under the [Parity Public License](https://paritylicense.com/versions/7.0.0). This license is not OSI approved and potentially problematic for downstream Zephyr adopters that may have closed source bits, as they would depend on acquiring a commercial license for being able to keep using the Zephyr documentation system. The [project's README](https://github.com/docleaf-labs/docleaf#license) makes this even more clear: "If you have a closed source project that you would like to document with Docleaf then you must purchase a commercial license"

This was an oversight from the reviewers of the original PR (myself included) to not look more closely at the license of Docleaf, as I guess (for me at least) the assumption was that it was a continuation/replacement for Breathe and hence would have the same license.

**Impact**

See above. The Parity Public License is implying that anyone "develop[ing], operat[ing], or analyz[ing] with this software [ie. Docleaf]" needs to make their software available under an open source license such as APL, MIT, ... ("a license that allows everything [the Parity Public License] does").
This is putting unwanted restrictions on how downstream adopters may be able to freely extend Zephyr for commercial / closed source purposes should they be interested in surfacing documentation for closed-source stuff.

I see a few paths going forward (assuming I understood the terms of the license properly, and that Docleaf sticks to it for the foreseeable future):

- switch back to Breathe. It's slower than Docleaf, and potentially unmaintained/less maintained going forward (?) but it's been doing the job until now.
- should Docleaf maintain some level of compatibility with Breathe, we might be able to provide both options? But this would need to be tested in CI and I am not sure it'd be desirable to systematically spend 2x the already long-ish time to build our docs in CI for each PR...

**Environment (please complete the following information):**

3.4.99 / 84e4ffcd56348f96ea532ece89f5eda9c62e8c56

**Additional context**

@michaeljones, feel free to add/comment here, especially if I misunderstood the terms of the license.

cc @nashif @gmarull

### PR DESCRIPTION
**Pull request by kartben**
This reverts the subset of commits from #59570 that introduced docleaf and fixes #61284.

- Revert "doc: Switch to using docleaf for doxygen entities"
- Revert "doc: Update requirements-doc.txt to use docleaf"
- Revert "doc: Replace instances of 'Breathe' with 'Docleaf'"
- Revert "doc: Adjust docleaf configuration as needed"
- Revert "doc: Remove :members: on doxygengroup directives"

Also added a commit marking some apparently unsupported and recently added constructs in `wifi.h` (namely, arrays with designated initializers) as INTERNAL_HIDDEN as they seem to be causing an issue (and are AFAICT meant to be internal anyway).

CI-built doc website: https://builds.zephyrproject.io/zephyr/pr/61339/docs/

### CODE REVIEW

**Comment by carlescufi** [comment]
@michaeljones please provide your feedback here

**Comment by kartben** [comment]
> @michaeljones please provide your feedback here

I think it's been provided in https://github.com/zephyrproject-rtos/zephyr/issues/61284#issuecomment-1669610304 already

**Comment by michaeljones** [comment]
Yep, seems to be a necessary step given the license. Sorry that it wasn't clearer from the start.

If there is the interest and the funding behind Zephyr for such things then I'd be happy to discuss a license exception that could also apply to Zephyr derivatives. That said, it sounds like you have a path forward exploring a Doxygen focused solution.

**Review by gmarull — DISMISSED** [review]

**Review by jhedberg — DISMISSED** [review]

**Comment by michaeljones** [comment]
@gmarull - There are two partially active maintainers on Breathe. I won't be actively working on it (and honestly haven't meaningfully in a long time) but they might. Development is slow and primarily responds to changes in Sphinx to my knowledge. They would benefit from funding if that is an option.

**Comment by kartben** [comment]
> Also added a commit marking some apparently unsupported and recently added constructs in `wifi.h` (namely, arrays with designated initializers) as INTERNAL_HIDDEN as they seem to be causing an issue (and are AFAICT meant to be internal anyway).

Mmm they're not that recent actually, so not sure why they are now failing (@gmarull @michaeljones any brilliant idea as to why this could be the case?). I guess I could add these as known warning but...

```
...
C:\Users\kartben\zephyrproject\zephyr\doc\_build\src\connectivity\networking\api\wifi.rst:30: WARNING: Invalid C declaration: Expected identifier in nested name. [error at 61]
  static const char *const wifi_twt_negotiation_type2str [] = {[WIFI_TWT_INDIVIDUAL] = "TWT individual negotiation",[WIFI_TWT_BROADCAST] = "TWT broadcast negotiation",[WIFI_TWT_WAKE_TBTT] = "TWT wake TBTT negotiation",}
...
```

Also, not sure they are meant to be internal either (although they might better off handled using helper functions just like other `*_txt(...)` helpers in this file?  /cc @krish2718 

Marking as DNM for now and will try to investigate more.

File: include/zephyr/net/wifi.h
Comment (gmarull) [inline include/zephyr/net/wifi.h:329]: actually... why do we have static arrays defined in headers? This should be in a C file...

Comment (kartben) [reply include/zephyr/net/wifi.h:329]: +1, I actually raised the issue to @krish2718 in this discussion https://github.com/zephyrproject-rtos/zephyr/pull/60686#discussion_r1289898763

Comment (krish2718) [reply include/zephyr/net/wifi.h:329]: This originally was an `inline` function but was later changed to `static arrays`, I can move to `wifi_shell.c` and as extern (As they might be used by others).

Comment (kartben) [reply include/zephyr/net/wifi.h:329]: If it's is desirable it can be used by others, and for consistency, would it make sense to turn it into a `wifi_ps_txt()` helper function, similar to how it's done for e.g. https://github.com/zephyrproject-rtos/zephyr/blob/4cbb9d8f5f464325948b41ac24939d4434f6ab8c/include/zephyr/net/wifi.h#L52-L75

Thanks!

Comment (krish2718) [reply include/zephyr/net/wifi.h:329]: `wifi_security_txt` this was the original code, and `wifi_ps_txt` was also added but due to a review comment went with static array.

Comment (kartben) [reply include/zephyr/net/wifi.h:329]: > So, will take care of converting this to static array and moving to C file.

Thanks! Please ping me when done so that I can update this PR accordingly.

Comment (krish2718) [reply include/zephyr/net/wifi.h:329]: Sure, but lets not wait for that PR to be ready soon :), it might be a while.

**Comment by kartben** [comment]
FWIW I am waiting for #60686 to be merged so that the change from 195c01be62012811aaade92e703066fed268c5dd can be taken care off there rather than in this PR.

**Comment by kartben** [comment]
@gmarull @jhedberg please revisit, I've now rebased to incorporate the latest changes in wifi.h that make breathe happy again.
CI built website: https://builds.zephyrproject.io/zephyr/pr/61339/docs/

**Review by gmarull — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Pull request merged**
